src/hostname/hostnamed.c

```diff
@@ -367,33 +367,53 @@ static int context_update_kernel_hostname(
         return r; /* 0 if no change, 1 if something was done  */
 }
 
+static void unset_statp(struct stat **p) {
+        if (!*p)
+                return;
+
+        **p = (struct stat) {};
+}
```

Comment:
This is a good fix. I wonder if we could make it a bit less verbose... Maybe just unconditionally unset the stat cache? This means we'll re-read the file on the next query, but that's no problem.